src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -139,23 +151,53 @@ class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig, protected
+    val machineConfig = MachineConfig(clusterRequest.machineConfig, clusterDefaultsConfig)
+
+    // Create a config for the master node, if properties are not specified in request, use defaults
+    val masterConfig = new InstanceGroupConfig()
+      .setMachineTypeUri(machineConfig.masterMachineType.get)
+      .setDiskConfig(new DiskConfig().setBootDiskSizeGb(machineConfig.masterDiskSize.get))
 
     // Create a Cluster Config and give it the GceClusterConfig, the NodeInitializationAction and the SoftwareConfig
-    val clusterConfig = new ClusterConfig()
-      .setGceClusterConfig(gce)
-      .setInitializationActions(initActions.asJava).setSoftwareConfig(softwareConfig)
+    createClusterConfig(machineConfig).setGceClusterConfig(gceClusterConfig)
+      .setInitializationActions(initActions.asJava)
+      .setMasterConfig(masterConfig)
+  }
 
-    // Create a Cluster and give it a name and a Cluster Config
-    val cluster = new DataprocCluster()
-      .setClusterName(clusterName.string)
-      .setConfig(clusterConfig)
+  private def createClusterConfig(machineConfig: MachineConfig): ClusterConfig = {
+    // If the number of workers is zero, make a Single Node cluster, else make a Standard one
+    if (machineConfig.numberOfWorkers.get == 0) {
+      // Create a SoftwareConfig and set a property that makes the cluster have only one node
+      val softwareConfig = new SoftwareConfig().setProperties(Map("dataproc:dataproc.allow.zero.workers" -> "true").asJava)
+      new ClusterConfig().setSoftwareConfig(softwareConfig)
+    }
+    else // Standard, multi node cluster
+      getStandardConfig(machineConfig)
+  }
 
-    // Create a dataproc create request and give it the google project, a zone, and the Cluster
-    val request = dataproc.projects().regions().clusters().create(googleProject.string, dataprocConfig.dataprocDefaultRegion, cluster)
+  private def getStandardConfig(machineConfig: MachineConfig): ClusterConfig = {
+    // Set the configs of the non-preemptible, primary worker nodes
+    val clusterConfigWithWorkerConfigs = new ClusterConfig().setWorkerConfig(getPrimaryWorkerConfig(machineConfig))
 
-    executeGoogleRequestAsync(googleProject, clusterName.toString, request)  // returns a Future[DataprocOperation]
+    // If the number of preemptible workers is greater than 0, set a secondary worker config
+    if (machineConfig.numberOfPreemptibleWorkers.get > 0) {
+      val preemptibleWorkerConfig = new InstanceGroupConfig()
+        .setIsPreemptible(true)
+        .setNumInstances(machineConfig.numberOfPreemptibleWorkers.get)
+
+      clusterConfigWithWorkerConfigs.setSecondaryWorkerConfig(preemptibleWorkerConfig)
+    } else clusterConfigWithWorkerConfigs
+  }
+
+  private def getPrimaryWorkerConfig(machineConfig: MachineConfig): InstanceGroupConfig = {
+    val workerDiskConfig = new DiskConfig()
+      .setBootDiskSizeGb(machineConfig.workerDiskSize.get)
```

Comment:
Instead of `.get` we could use `.orNull` which is equivalent to `.getOrElse(null)`. We expect all values to be present; but the Java API should be able to deal with null just in case.

Reply:
In that case, if we have a surprise `None` here then we'd use Google's default value, which could someday differ from our config-file defaults

Reply:
Verdict? - Do we want to use .orNull and allow Google to set the defaults or is it better to just fail in this case? I would think failing would be preferable?

Reply:
Also when I try to use orNull, I get `type arguments [Integer] do not conform to method orNull's type parameter bounds [A1 >: Int]` :(

Reply:
Yeah on second thought, maybe failing would be better than surprise Google defaults.

The type error is related to converting from scala `Int` to Java `Integer`. There's probably a way around it, but I think this is fine. :)